### DISCUSSION TITLE
2.0: inconsistent behavior of `default` and `default_factory` for `mapped_column`

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [sqlalchemy/sqlalchemy#9877] [Usage Questions] [answered]
Take a look at this MRE, see comments to `print`s
`sqlalchemy==2.0.15`

```python
from uuid import UUID, uuid4

from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column


class Base(DeclarativeBase, MappedAsDataclass):
    pass


class Test(Base):
    __tablename__ = 'test'

    id: Mapped[UUID] = mapped_column(primary_key=True)
    flag_df: Mapped[bool] = mapped_column(default_factory=lambda: True, init=False)
    flag_d: Mapped[bool] = mapped_column(default=True, init=False)


test = Test(id=uuid4())

print(f'flag from default factory = {test.flag_df}')  # shows True. Behavior is OK
print(f'flag from default = {test.flag_d}')  # shows None. Most likely it is a bug
```

**Comment by zzzeek** [comment]
seems likely to be a bug however I have no idea what dataclasses is doing here such that those defaults get set.  We call the dataclasses original `__init__()` method.   I have no idea how to fix, if you want to take a look.

**Comment by AlexanderPodorov** [comment]
Yes, using the vanilla dataclasses, it works:
```python
from dataclasses import dataclass, field

@dataclass
class Test:
    f: bool = field(default=True, init=False)

test = Test()

print(f'value = {test.f}')  # returns True
```
So technically it's a bug, sqlalchemy is not compliant with dataclasses in that regard.
I can try to take a look, however not sure I will find out anything.

**Comment by CaselIT** [comment]
I've created an issue https://github.com/sqlalchemy/sqlalchemy/issues/9879

**Comment by CaselIT** [comment]
Ok this is not good. Dataclass uses this pattern for default values

```py
class Foo:
  value = 'class_default'
```
doing this
```py
from dataclasses import dataclass, field

@dataclass
class Test:
    f: bool = field(default=True, init=False)
    fd: bool = field(default_factory=lambda: False, init=False)

test = Test()
print(test.__dict__)
```

prints `{'fd': False}`: there is no `f` in the class dict.

@zzzeek is the fix here to transform all defaults into default_factories?

**Comment by zzzeek** [comment]
Oh haha is that what they do?  Ha right.  What do they do when the class has slots ? anyway yeah we make default factories when we see a default and init is false.  Great idea .

**Comment by zzzeek** [comment]
Or, maybe we raise an error there and say to use a default factory directly.

**Comment by zzzeek** [comment]
let's just try going with "we'll change to a default factory and pretend everything worked".  if it causes problems someday, then we'll have a good case study on why we really need to never assume anything and we can deprecate it.

**Reply by CaselIT** [reply]
makes sense

**Comment by AlexanderPodorov** [comment]
Thanks guys for the investigation. I’m just curious why `test.f` returns the correct value in case of vanilla data classes, but returns `None` for `sqlalchemy` synthesized data class. It looks like something is missing here, but I understand it could be fairly hard to figure out.
I personally don’t think it would be good to deprecate `default` attribute, because it will break the idea about dataclass-like behavior and probably could cause issues for type checkers.
I think the suggested approach with wrapping default to default_factory is ideal. Since those attributes should be mutually exclusive.
Thanks!

**Reply by zzzeek** [reply]
note this is only when init=False.    it works with plain dataclasses because it's setting that value as a class variable, which is not compatible with descriptors nor with slots (so not sure what dataclasses does if slots is set up).

if we raised for "default=X, init=False" that wouldn't break for type checkers, it would break at runtime, you'd replace with with "default_factory = lambda: X, init=False", then the code would work.   but as I said we'd likely just transparently change this assuming there's no typing issue.

**Comment by AlexanderPodorov** [comment]
As far as I remember 1.4 style approach with data classes [here](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#mapping-pre-existing-dataclasses-using-declarative-style-fields) also had this behavior.

**Comment by AlexanderPodorov** [comment, marked as answer]
Issue already addressed in https://github.com/sqlalchemy/sqlalchemy/issues/9879. Discussion is closed
